src/workerd/io/actor-sqlite.c++

```diff
@@ -157,11 +158,68 @@ void ActorSqlite::onWrite() {
       // rather than after the callback.
       { auto drop = kj::mv(txn); }
 
-      return commitCallback();
+      return commitImpl();
```

Comment:
Looking at this I realize that the pre-commit alarm update actually needs to happen not just before calling the commit callback, but actually before committing the transaction. The alarm update should become part of the transaction.

(Technically the implicit transaction is only important for correctness in workerd -- in production, SRS provides another layer of safety via the commit callback. But we probably should make this correct for workerd. Also, lumping more stuff into fewer transactions performs better, as sqlite is then able to coalesce writes to the same pages.)

Reply:
When you say the pre-commit alarm update needs to happen before committing the transaction, do you mean that we need to asynchronously block for any necessary alarm notification change requests to complete just prior to the above `txn->commit()` call?

If so, does that put us in a situation where we're holding a sqlite transaction open while waiting for async work to complete, like Justin mentioned on the spec?  Or is there a way to avoid that?

Reply:
Yes, that's what I meant.

It should be OK to hold it open. It's already being held open until the `kj::evalLater` above resolves -- arbitrary async stuff could have happened in the meantime. It's possible that more JavaScript code will run, even, in which case changes made in that code will be merged into the same implicit transaction. Again, that'll all fine: the point of the implicit transaction is only to make sure that two queries executed in sequence without an await are applied atomically, but there's no reason that it _can't_ extend that atomicity further, such that it actually crosses an await.

There is some complexity if the application requests an explicit asynchronous transaction in the meantime. When that happens we normally commit any implicit transaction immediately -- see the constructor of ExplicitTransaction. Perhaps the way we handle this is: Before committing the implicit transaction to make way for an explicit one, if an alarm update is needed, we at least _start_ the `setAlarm()` call, but we commit the transaction without waiting for it. In workerd, this works because `setAlarm()` updates the alarm synchronously, so by the time we commit the transaction, the alarm has in fact been updated. Whereas in production, we rely on SRS commits as the true guardian of atomicity, and the SRS commit won't be sent until after the alarm has successfully udpated.

Reply:
Oh... I had been assuming that the `commitCallback()` had to return for the change to be considered durably persisted, but I guess that's only true in production, and that the workerd local database state is effectively also the workerd durable state.

In which case, do we theoretically need similar logic in the explicit transaction's `commit()` -- i.e. starting any necessary alarm notification request prior to the root transaction's `run("RELEASE _cf_savepoint_"...)`?